### PR TITLE
Fix k++

### PR DESCRIPTION
**Pull request by amrukwa**
It still doesn't work and I can't see the issue.

### CODE REVIEW

File: initialization.py [hunk 1]
```diff
@@ -23,9 +22,20 @@ def kpp_init(data, n_clusters):
...
+    dist = fix_the_distance(x, centroids, 'correlation')
     p_dist = np.amin(dist, axis=1)
+    p_dist = p_dist**2
     all_dist = np.sum(p_dist)
     probability = [(p_dist[i]) / all_dist for i in range(datasize)]
     new_centre = x[np.random.choice(datasize, 1, p=probability)]
     return new_centre
+
+
+def fix_the_distance(x, centroids, dist_metric):
```
Comment (gmrukwa) [inline initialization.py:34]: Location of this function in `initialization.py` is counter-intuitive.

File: clusters_n.py [hunk 2]
```diff
@@ -1,19 +1,17 @@
...
-import scipy.spatial.distance as ssdist
 import numpy as np
+import initialization as init
 
 
 def _dunn_index(x, estimator):
     datasize = x.shape[0]
-    intercluster_distance = ssdist.cdist(estimator.centroids, estimator.centroids, metric='correlation')
-    intercluster_distance[np.isnan(intercluster_distance)] = 0
+    intercluster_distance = init.fix_the_distance(estimator.centroids, estimator.centroids, 'correlation')
```
Comment (gmrukwa) [inline clusters_n.py:8]: You hardcoded the `correlation` metric.

File: clusters_n.py [hunk 3]
```diff
@@ -1,19 +1,17 @@
...
     min_ctc = intercluster_distance[1, 0]
     for i in range(estimator.n_clusters):
         for j in range(estimator.n_clusters):
             if min_ctc > intercluster_distance[i, j] > 0:
                 min_ctc = intercluster_distance[i, j]
-    distance_ctp = ssdist.cdist(x, estimator.centroids, metric='correlation')
-    distance_ctp[np.isnan(distance_ctp)] = 0
+    distance_ctp = init.fix_the_distance(x, estimator.centroids, 'correlation')
```
Comment (gmrukwa) [inline clusters_n.py:14]: You hardcoded the `correlation` metric.

File: clusters_n.py [hunk 4]
```diff
@@ -1,19 +1,17 @@
...
     intracluster_distance = [distance_ctp[estimator.labels_[_]] for _ in range(datasize)]
```
Comment (gmrukwa) [inline clusters_n.py:15]: `_` should only be used when its value is ignored.

File: initialization.py [hunk 5]
```diff
@@ -1,9 +1,8 @@
 import numpy as np
-import kmeans_functions as kmf
+import scipy.spatial.distance as ssdist
 
 
 def random_init(data, n_clusters):
```
Comment (gmrukwa) [inline initialization.py:5]: This function should operate accepting `random_state`, just like `sklearn` estimators. See [this](https://github.com/scikit-learn/scikit-learn/blob/95d4f0841/sklearn/cluster/_kmeans.py#L843), [this](https://github.com/scikit-learn/scikit-learn/blob/95d4f0841d57e8b5f6b2a570312e9d832e69debc/sklearn/utils/validation.py#L800) and [this](https://github.com/scikit-learn/scikit-learn/blob/95d4f0841/sklearn/cluster/_kmeans.py#L926). Alternatively, you can try [another approach](https://github.com/gmrukwa/divik/blob/master/divik/core/_seed.py#L17).

File: initialization.py [hunk 6]
```diff
@@ -23,9 +22,20 @@ def kpp_init(data, n_clusters):
 
 def _do_for_kpp(centroids, x):
```
Comment (gmrukwa) [inline initialization.py:23]: Unclear function name `_do_for_kpp`.

File: initialization.py (hunk 1 above)
Comment (gmrukwa) [inline initialization.py:34]: Omit `the` in function names.

File: initialization.py [hunk 7]
```diff
@@ -23,9 +22,20 @@ def kpp_init(data, n_clusters):
...
     return new_centre
+
+
+def fix_the_distance(x, centroids, dist_metric):
+    distance = ssdist.cdist(x, centroids, metric=dist_metric)  # here is sth to change
+    distance[np.isnan(distance)] = 0
+    for i in range(x.shape[0]):
+        for j in range(centroids.shape[0]):
+            if (x[i] == centroids[j]).all():
+                distance[i, j] = 0
```
Comment (gmrukwa) [inline initialization.py:40]: This should already be guaranteed by `cdist`, why do you need that?

File: initialization.py [hunk 8]
```diff
@@ -23,9 +22,20 @@ def kpp_init(data, n_clusters):
...
     p_dist = np.amin(dist, axis=1)
+    p_dist = p_dist**2
     all_dist = np.sum(p_dist)
     probability = [(p_dist[i]) / all_dist for i in range(datasize)]
     new_centre = x[np.random.choice(datasize, 1, p=probability)]
     return new_centre
+
+
+def fix_the_distance(x, centroids, dist_metric):
+    distance = ssdist.cdist(x, centroids, metric=dist_metric)  # here is sth to change
```
Comment (gmrukwa) [inline initialization.py:35]: This function **computes** the distances, doesn't only **fix**.

File: reducing_dims.py [hunk 9]
```diff
@@ -1,6 +1,5 @@
 import numpy as np
 from scipy.linalg import orth
-import pandas as pd
 np.seterr(divide='ignore', invalid='ignore')
```
Comment (gmrukwa) [inline reducing_dims.py:3]: Why do you need that? This kind of error indicates that you have broken assumptions somewhere in your code, what's the point of silencing it? You cannot find the error if you silenced it O.o

File: test/test_kmeans.py [hunk 10]
```diff
@@ -15,26 +15,19 @@ def test_fails_for_repeating_centroids():
     data = pd.read_csv('/Users/amruk/PycharmProjects/kmeans/setA.txt', delimiter=' ')
     file.close()
     data1 = data[["Gsto1", "Gstm1", "Cd9", "Prdx1", "Coro1a", "Rac2", "Perp"]].values
-    for i in range(10):
+    i = 2
+    while i <= 10:
         centroids = kmf.initialization(i, 'k-means++', data1)
         essence = np.unique(centroids, axis=0)
         assert essence.shape[0] == centroids.shape[0]
+        i +=1
```
Comment (gmrukwa) [inline test/test_kmeans.py:23]:
```suggestion
    for i in range(2, 11):
        centroids = kmf.initialization(i, 'k-means++', data1)
        essence = np.unique(centroids, axis=0)
        assert essence.shape[0] == centroids.shape[0]
```

File: test/test_kmeans.py [hunk 11]
```diff
@@ -15,26 +15,19 @@ def test_fails_for_repeating_centroids():
     data = pd.read_csv('/Users/amruk/PycharmProjects/kmeans/setA.txt', delimiter=' ')
```
Comment (gmrukwa) [inline test/test_kmeans.py:15]: Do not use real data in tests. Especially that you are using absolute path - this will not work anywhere but on your PC.
You have [`make_blobs`](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_blobs.html) and [other functions](https://scikit-learn.org/stable/modules/classes.html#samples-generator) for that.

File: test/test_kmeans.py [hunk 12]
```diff
@@ -15,26 +15,19 @@ def test_fails_for_repeating_centroids():
     data = pd.read_csv('/Users/amruk/PycharmProjects/kmeans/setA.txt', delimiter=' ')
     file.close()
     data1 = data[["Gsto1", "Gstm1", "Cd9", "Prdx1", "Coro1a", "Rac2", "Perp"]].values
-    for i in range(10):
+    i = 2
+    while i <= 10:
```
Comment (gmrukwa) [inline test/test_kmeans.py:19]: Don't make it a loop here. If it fails, how do you know what were the conditions of failure? Was it for 5 or 8 clusters? Rather use [parameterized tests](https://docs.pytest.org/en/latest/parametrize.html), so the test framework keeps track of everything for you.

File: initialization.py (hunk 7 above)
Comment (amrukwa) [reply initialization.py:40]: I checked and well, the correlation function doesn't return 0 for the distance of vector to itself; or I have some strange bug elsewhere, but it resulted in empty slices

File: kmeans_functions.py [hunk 13]
```diff
@@ -37,18 +38,29 @@ def _centroid_by_means(n_clusters, centroids, labels, x):
...
+
+def check_random_state(seed):
+    if seed is None or seed is np.random:
+        return np.random.mtrand._rand
+    if isinstance(seed, numbers.Integral):
+        return np.random.RandomState(seed)
+    if isinstance(seed, np.random.RandomState):
+        return seed
+    raise ValueError('%r cannot be used to seed a numpy.random.RandomState'
+                     ' instance' % seed)
```
Comment (gmrukwa) [inline kmeans_functions.py:66]: Missing newline at the end of file.

File: kmeans_functions.py [hunk 14]
```diff
@@ -37,18 +38,29 @@ def _centroid_by_means(n_clusters, centroids, labels, x):
...
-                distance[i, j] = 0
+    # distance[np.isnan(distance)] = 0
+    # for i in range(x.shape[0]):
+    #     for j in range(centroids.shape[0]):
+    #         if (x[i] == centroids[j]).all():
+    #             distance[i, j] = 0
     return distance
+
+
+def check_random_state(seed):
```
Comment (gmrukwa) [inline kmeans_functions.py:58]: You could import that function from `scikit-learn`, couldn't you?

File: kmeans_functions.py [hunk 15]
```diff
@@ -37,18 +38,29 @@ def _centroid_by_means(n_clusters, centroids, labels, x):
...
-    distance[np.isnan(distance)] = 0
-    for i in range(x.shape[0]):
-        for j in range(centroids.shape[0]):
-            if (x[i] == centroids[j]).all():
-                distance[i, j] = 0
+    # distance[np.isnan(distance)] = 0
+    # for i in range(x.shape[0]):
+    #     for j in range(centroids.shape[0]):
+    #         if (x[i] == centroids[j]).all():
+    #             distance[i, j] = 0
```
Comment (gmrukwa) [inline kmeans_functions.py:54]: Delete commented code.

File: kmeans_functions.py [hunk 16]
```diff
@@ -37,18 +38,29 @@ def _centroid_by_means(n_clusters, centroids, labels, x):
...
     datasize = x.shape[0]
-    distance = fix_the_distance(x, centroids, 'correlation')
-    inertia = np.sum([(distance[_]) ** 2 for _ in range(datasize)])
+    distance = compute_distance(x, centroids, metric)
+    inertia = np.sum([(distance[i]) ** 2 for i in range(datasize)])
     return inertia
 
 
-def fix_the_distance(x, centroids, dist_metric):
+def compute_distance(x, centroids, dist_metric):
```
Comment (gmrukwa) [inline kmeans_functions.py:48]: Why do you need this function at all? It just calls another function and returns its result.

File: kmeans_functions.py [hunk 17]
```diff
@@ -1,31 +1,32 @@
...
-def compute_centroids(n_clusters, centroids, labels, n_iter, max_iter, x):
+def compute_centroids(n_clusters, centroids, labels, n_iter, max_iter, x, metric):
     while n_iter < max_iter:
         if n_iter > 0:
             prev_centroids = centroids
         centroids = _centroid_by_means(n_clusters, centroids, labels, x)
-        labels = labeling(centroids, x)
-        if n_iter > 5 and np.all(centroids == prev_centroids):
+        labels = labeling(centroids, x, metric)
+        if n_iter > 0 and np.all(centroids == prev_centroids):
```
Comment (gmrukwa) [inline kmeans_functions.py:29]: It's better to compare previous labels - less numbers to compare. Additionally comparison would be done on `int` instead of `float`.
